File: book/snippets/python/src/hello.py
# ANCHOR: full
import asyncio
from nostr_sdk import Keys, Client, EventBuilder, NostrSigner


async def hello():
    # ANCHOR: client
    keys = Keys.generate()
    signer = NostrSigner.keys(keys)
    client = Client(signer)
    # ANCHOR_END: client

    # ANCHOR: connect
    await client.add_relay("wss://relay.damus.io")
    await client.connect()
    # ANCHOR_END: connect

    # ANCHOR: publish
    builder = EventBuilder.text_note("Hello, rust-nostr!")
    output = await client.send_event_builder(builder)
    # ANCHOR_END: publish

    # ANCHOR: output
    print(f"Event ID: {output.id.to_bech32()}")
    print(f"Sent to: {output.success}")
    print(f"Not send to: {output.failed}")
    # ANCHOR_END: output

if __name__ == '__main__':
   asyncio.run(hello())
# ANCHOR_END: full

File: book/snippets/rust/src/hello.rs
// ANCHOR: full
use nostr_sdk::prelude::*;

pub async fn hello() -> Result<()> {
    // ANCHOR: client
    let keys: Keys = Keys::generate();
    let client = Client::new(keys);
    // ANCHOR_END: client

    // ANCHOR: connect
    client.add_relay("wss://relay.damus.io").await?;
    client.connect().await;
    // ANCHOR_END: connect

    // ANCHOR: publish
    let builder = EventBuilder::text_note("Hello, rust-nostr!");
    let output = client.send_event_builder(builder).await?;
    // ANCHOR_END: publish

    // ANCHOR: output
    println!("Event ID: {}", output.id().to_bech32()?);
    println!("Sent to: {:?}", output.success);
    println!("Not sent to: {:?}", output.failed);
    // ANCHOR_END: output

    Ok(())
}

// ANCHOR_END: full
